llvm/lib/Target/SBF/SBFInstrInfo.cpp

```diff
@@ -104,94 +141,262 @@ bool SBFInstrInfo::analyzeBranch(MachineBasicBlock &MBB,
+
+      if (isCondBranchOpcode(LastOpc)) {
+        // Block ends with fall-through condbranch.
+        parseCondBranch(LastInst, TBB, Cond);
+        return false;
+      }
+      return true; // Can't handle indirect branch.
     }
-    // Cannot handle conditional branches
+    SecondLastInst = &*I;
+    SecondLastOpc = SecondLastInst->getOpcode();
+  }
+
+  // If there are three terminators, we don't know what sort of block this is.
+  if (SecondLastInst && I != MBB.begin() && isUnpredicatedTerminator(*--I))
     return true;
+
+  // If the block ends with a conditional jump and a JA, handle it.
+  if (isCondBranchOpcode(SecondLastOpc) && isUncondBranchOpcode(LastOpc)) {
+    parseCondBranch(SecondLastInst, TBB, Cond);
+    FBB = LastInst->getOperand(0).getMBB();
+    return false;
   }
 
-  return false;
+  // If the block ends with two unconditional branches, handle it.  The second
+  // one is not executed, so remove it.
+  if (isUncondBranchOpcode(SecondLastOpc) && isUncondBranchOpcode(LastOpc)) {
+    TBB = SecondLastInst->getOperand(0).getMBB();
+    I = LastInst;
+    if (AllowModify)
+      I->eraseFromParent();
+    return false;
+  }
+
+  // Otherwise, can't handle this.
+  return true;
 }
 
 unsigned SBFInstrInfo::insertBranch(MachineBasicBlock &MBB,
                                     MachineBasicBlock *TBB,
                                     MachineBasicBlock *FBB,
                                     ArrayRef<MachineOperand> Cond,
-                                    const DebugLoc &DL,
-                                    int *BytesAdded) const {
+                                    const DebugLoc &DL, int *BytesAdded) const {
   assert(!BytesAdded && "code size not handled");
 
   // Shouldn't be a fall through.
   assert(TBB && "insertBranch must not be told to insert a fallthrough");
 
+  if (BytesAdded)
+    *BytesAdded = 8;
+
   if (Cond.empty()) {
     // Unconditional branch
     assert(!FBB && "Unconditional branch with multiple successors!");
     BuildMI(&MBB, DL, get(SBF::JMP)).addMBB(TBB);
     return 1;
   }
 
-  llvm_unreachable("Unexpected conditional branch");
+  // See the order we parse the jump information in `parseCondBranch`
+  BuildMI(&MBB, DL, get(Cond[0].getImm()))
+      .add(Cond[1])
+      .add(Cond[2])
+      .addMBB(TBB);
+
+  if (FBB) {
+    BuildMI(&MBB, DL, get(SBF::JMP)).addMBB(FBB);
+    if (BytesAdded)
+      *BytesAdded += 8;
+  }
+
+  return 1;
 }
 
 unsigned SBFInstrInfo::removeBranch(MachineBasicBlock &MBB,
                                     int *BytesRemoved) const {
   assert(!BytesRemoved && "code size not handled");
 
-  MachineBasicBlock::iterator I = MBB.end();
-  unsigned Count = 0;
-
-  while (I != MBB.begin()) {
-    --I;
-    if (I->isDebugInstr())
-      continue;
-    if (I->getOpcode() != SBF::JMP)
-      break;
-    // Remove the branch.
-    I->eraseFromParent();
-    I = MBB.end();
-    ++Count;
+  MachineBasicBlock::iterator I = MBB.getLastNonDebugInstr();
```

Comment:
This one is from here:

https://github.com/anza-xyz/llvm-project/blob/04d1c6c27ed5001b34fea10e31e0a1a87ebfd59e/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp#L540